Lazy load libcurl, allowing for an SSL/TLS backend-specific libcurl

dscho opened the pull request:
As per https://github.com/git-for-windows/git/issues/4350#issuecomment-1485041503, the major block for upgrading Git for Windows' OpenSSL from v1.1 to v3 is the tricky part where such an upgrade would break `git fetch`/`git clone` and `git push` because the libcurl depends on the OpenSSL DLL, and the major version bump will _change_ the file name of said DLL.

To overcome that, the plan is to build libcurl flavors for each supported SSL/TLS backend, aligning with the way MSYS2 builds libcurl, then switch Git for Windows' SDK to the Secure Channel-flavored libcurl, and teach Git to look for the specific flavor of libcurl corresponding to the `http.sslBackend` setting (if that was configured).

Here is the PR to teach Git that trick.

rimrul commented on config.mak.uname:
```diff
@@ -670,6 +670,7 @@ ifeq ($(uname_S),MINGW)
 	DEFAULT_HELP_FORMAT = html
 	HAVE_PLATFORM_PROCINFO = YesPlease
 	CSPRNG_METHOD = rtlgenrandom
+	LAZY_LOAD_LIBCURL = YesDoThatPlease
```
It might make sense to mention in the commit message how/why this will help with our mainzenance.

dscho replied:
Sure, I can see that 😁

dscho commented on Makefile:
```diff
@@ -1603,10 +1608,17 @@ else
+	ifdef LAZY_LOAD_LIBCURL
+		LAZY_LOAD_LIBCURL_OBJ = compat/lazy-load-curl.o
+		OBJECTS += $(LAZY_LOAD_LIBCURL_OBJ)
+		CURL_CFLAGS = -DCURL_STATICLIB
```
Right, this is _required_ so that the `curl.h` headers do not declare the functions as `dllimport`.

dscho commented on Makefile:
```diff
@@ -1603,10 +1608,17 @@ else
-	ifndef CURL_LDFLAGS
-		CURL_LDFLAGS = $(eval CURL_LDFLAGS := $$(shell $$(CURL_CONFIG) --libs))$(CURL_LDFLAGS)
+	ifdef LAZY_LOAD_LIBCURL
+		LAZY_LOAD_LIBCURL_OBJ = compat/lazy-load-curl.o
```
Good idea!

dscho commented on compat/lazy-load-curl.c:
```diff
@@ -0,0 +1,337 @@
+#define LIBCURL_FILE_NAME(base) base ".4.dylib"
+#else
+#define LIBCURL_FILE_NAME(base) base ".so.4"
+#endif
```
Hmm. I just had a look and git-for-windows/git-sdk-32@d85efcecef95b40321c754ee3ed1dcd39838541d suggests that even for versions as old as v7.49, the number was a `4`...

dscho replied:
Actually, I don't think we're in _any_ kind of danger here. The `4` is the [libcurl ABI version](https://curl.se/libcurl/abi.html) which has not changed since v7.16.0, i.e. October 2006...

rimrul replied:
It's not that simple. It shouldn't be a problem on Windows, but we might (unlikely but possible) encounter the 5 on other platforms.

See

https://github.com/curl/curl/commit/3f3d6ebe665f32f3fe6c3c49df86bdffd223090c

https://github.com/curl/curl/pull/1861

It's unlikely that we'll encounter this in the wild and we could address it when/if it happens.

dscho replied:
True. I added a code comment to help with this, should it become a problem.

dscho commented on compat/lazy-load-curl.c:
```diff
@@ -1,9 +1,12 @@
 #include "../git-curl-compat.h"
+#ifndef WIN32
 #include <dlfcn.h>
+#endif
```
Good point.

dscho commented:
@rimrul I believe that I have addressed all of your feedback so far.

<details><summary>Range-diff</summary>



* 1:  2d1ad108a9c4 ! 1:  55b28f9b965e http: optionally load libcurl lazily

   ``````diff
   @@ Makefile: include shared.mak
     #
     #     CURL_LDFLAGS=-lcurl
     #
   -+# Define LAZY_LOAD_LIBCURL to dynamically load the libcurl; This can be useful
   ++# Define LAZYLOAD_LIBCURL to dynamically load the libcurl; This can be useful
    +# if Multiple libcurl versions exist (with different file names) that link to
    +# various SSL/TLS backends, to support the `http.sslBackend` runtime switch in
    +# such a scenario.
   @@ Makefile: else
     
    -	ifndef CURL_LDFLAGS
    -		CURL_LDFLAGS = $(eval CURL_LDFLAGS := $$(shell $$(CURL_CONFIG) --libs))$(CURL_LDFLAGS)
   -+	ifdef LAZY_LOAD_LIBCURL
   -+		LAZY_LOAD_LIBCURL_OBJ = compat/lazy-load-curl.o
   -+		OBJECTS += $(LAZY_LOAD_LIBCURL_OBJ)
   ++	ifdef LAZYLOAD_LIBCURL
   ++		LAZYLOAD_LIBCURL_OBJ = compat/lazyload-curl.o
   ++		OBJECTS += $(LAZYLOAD_LIBCURL_OBJ)
   ++		# The `CURL_STATICLIB` constant must be defined to avoid seeing the functions
   ++		# declared as DLL imports
    +		CURL_CFLAGS = -DCURL_STATICLIB
    +		CURL_LIBCURL = -ldl
    +	else
   @@ Makefile: else
     	ifdef USE_CURL_FOR_IMAP_SEND
     		BASIC_CFLAGS += -DUSE_CURL_FOR_IMAP_SEND
    -		IMAP_SEND_BUILDDEPS = http.o
   -+		IMAP_SEND_BUILDDEPS = http.o $(LAZY_LOAD_LIBCURL_OBJ)
   ++		IMAP_SEND_BUILDDEPS = http.o $(LAZYLOAD_LIBCURL_OBJ)
     		IMAP_SEND_LDFLAGS += $(CURL_LIBCURL)
     	endif
     	ifndef NO_EXPAT
   @@ Makefile: git-imap-send$X: imap-send.o $(IMAP_SEND_BUILDDEPS) GIT-LDFLAGS $(GITL
     		$(IMAP_SEND_LDFLAGS) $(LIBS)
     
    -git-http-fetch$X: http.o http-walker.o http-fetch.o GIT-LDFLAGS $(GITLIBS)
   -+git-http-fetch$X: http.o http-walker.o http-fetch.o $(LAZY_LOAD_LIBCURL_OBJ) GIT-LDFLAGS $(GITLIBS)
   ++git-http-fetch$X: http.o http-walker.o http-fetch.o $(LAZYLOAD_LIBCURL_OBJ) GIT-LDFLAGS $(GITLIBS)
     	$(QUIET_LINK)$(CC) $(ALL_CFLAGS) -o $@ $(ALL_LDFLAGS) $(filter %.o,$^) \
     		$(CURL_LIBCURL) $(LIBS)
    -git-http-push$X: http.o http-push.o GIT-LDFLAGS $(GITLIBS)
   -+git-http-push$X: http.o http-push.o $(LAZY_LOAD_LIBCURL_OBJ) GIT-LDFLAGS $(GITLIBS)
   ++git-http-push$X: http.o http-push.o $(LAZYLOAD_LIBCURL_OBJ) GIT-LDFLAGS $(GITLIBS)
     	$(QUIET_LINK)$(CC) $(ALL_CFLAGS) -o $@ $(ALL_LDFLAGS) $(filter %.o,$^) \
     		$(CURL_LIBCURL) $(EXPAT_LIBEXPAT) $(LIBS)
     
   @@ Makefile: $(REMOTE_CURL_ALIASES): $(REMOTE_CURL_PRIMARY)
     	cp $< $@
     
    -$(REMOTE_CURL_PRIMARY): remote-curl.o http.o http-walker.o GIT-LDFLAGS $(GITLIBS)
   -+$(REMOTE_CURL_PRIMARY): remote-curl.o http.o http-walker.o $(LAZY_LOAD_LIBCURL_OBJ) GIT-LDFLAGS $(GITLIBS)
   ++$(REMOTE_CURL_PRIMARY): remote-curl.o http.o http-walker.o $(LAZYLOAD_LIBCURL_OBJ) GIT-LDFLAGS $(GITLIBS)
     	$(QUIET_LINK)$(CC) $(ALL_CFLAGS) -o $@ $(ALL_LDFLAGS) $(filter %.o,$^) \
     		$(CURL_LIBCURL) $(EXPAT_LIBEXPAT) $(LIBS)
     
    
   - ## compat/lazy-load-curl.c (new) ##
   + ## compat/lazyload-curl.c (new) ##
    @@
    +#include "../git-compat-util.h"
    +#include "../git-curl-compat.h"
    +#include <dlfcn.h>
    +
   ++/*
   ++ * The ABI version of libcurl is encoded in its shared libraries' file names.
   ++ * This ABI version has not changed since October 2006 and is unlikely to be
   ++ * changed in the future. See https://curl.se/libcurl/abi.html for details.
   ++ */
   ++#define LIBCURL_ABI_VERSION 4
   ++
    +typedef void (*func_t)(void);
    +
    +#ifdef __APPLE__
   -+#define LIBCURL_FILE_NAME(base) base ".4.dylib"
   ++#define LIBCURL_FILE_NAME(base) base "." #LIBCURL_ABI_VERSION ".dylib"
    +#else
   -+#define LIBCURL_FILE_NAME(base) base ".so.4"
   ++#define LIBCURL_FILE_NAME(base) base ".so." #LIBCURL_ABI_VERSION
    +#endif
    +
    +static void *load_library(const char *name)
   ``````

* 2:  f1e357235227 ! 2:  b658d20b9628 http: support lazy-loading libcurl also on Windows

   ``````diff
   @@ Commit message
        This implements the Windows-specific support code, because everything is
        slightly different on Windows, even loading shared libraries.
    
   +    Note: I specifically do _not_ use the code from
   +    `compat/win32/lazyload.h` here because that code is optimized for
   +    loading individual functions from various system DLLs, while we
   +    specifically want to load _many_ functions from _one_ DLL here, and
   +    distinctly not a system DLL (we expect libcurl to be located outside
   +    `C:\Windows\system32`, something `INIT_PROC_ADDR` refuses to work with).
   +    Also, the `curl_easy_getinfo()`/`curl_easy_setopt()` functions are
   +    declared as vararg functions, which `lazyload.h` cannot handle. Finally,
   +    we are about to optionally override the exact file name that is to be
   +    loaded, which is a goal contrary to `lazyload.h`'s design.
   +
        Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>
    
     ## Makefile ##
    @@ Makefile: else
   - 		LAZY_LOAD_LIBCURL_OBJ = compat/lazy-load-curl.o
   - 		OBJECTS += $(LAZY_LOAD_LIBCURL_OBJ)
   + 		# The `CURL_STATICLIB` constant must be defined to avoid seeing the functions
   + 		# declared as DLL imports
     		CURL_CFLAGS = -DCURL_STATICLIB
    +ifneq ($(uname_S),MINGW)
    +ifneq ($(uname_S),Windows)
   @@ Makefile: else
     		ifndef CURL_LDFLAGS
     			CURL_LDFLAGS = $(eval CURL_LDFLAGS := $$(shell $$(CURL_CONFIG) --libs))$(CURL_LDFLAGS)
    
   - ## compat/lazy-load-curl.c ##
   + ## compat/lazyload-curl.c ##
    @@
     #include "../git-compat-util.h"
     #include "../git-curl-compat.h"
   @@ compat/lazy-load-curl.c
     #include <dlfcn.h>
    +#endif
     
   + /*
   +  * The ABI version of libcurl is encoded in its shared libraries' file names.
   +@@
   + 
     typedef void (*func_t)(void);
     
    +#ifndef WIN32
     #ifdef __APPLE__
   - #define LIBCURL_FILE_NAME(base) base ".4.dylib"
   + #define LIBCURL_FILE_NAME(base) base "." #LIBCURL_ABI_VERSION ".dylib"
     #else
   -@@ compat/lazy-load-curl.c: static func_t load_function(void *handle, const char *name)
   +@@ compat/lazyload-curl.c: static func_t load_function(void *handle, const char *name)
     	*(void **)&f = dlsym(handle, name);
     	return f;
     }
    +#else
   -+#define LIBCURL_FILE_NAME(base) base "-4.dll"
   ++#define LIBCURL_FILE_NAME(base) base "-" #LIBCURL_ABI_VERSION ".dll"
    +
    +static void *load_library(const char *name)
    +{
   ``````

* 3:  82baabb599ed ! 3:  2beaffb5c8fe http: when loading libcurl lazily, allow for multiple SSL backends

   ``````diff
   @@ Commit message
    
        Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>
    
   - ## compat/lazy-load-curl.c ##
   -@@ compat/lazy-load-curl.c: static curl_easy_setopt_pointer_type curl_easy_setopt_pointer_func;
   + ## compat/lazyload-curl.c ##
   +@@ compat/lazyload-curl.c: static curl_easy_setopt_pointer_type curl_easy_setopt_pointer_func;
     typedef CURLcode (*curl_easy_setopt_off_t_type)(CURL *curl, CURLoption opt, curl_off_t value);
     static curl_easy_setopt_off_t_type curl_easy_setopt_off_t_func;
     
   @@ compat/lazy-load-curl.c: static curl_easy_setopt_pointer_type curl_easy_setopt_p
     	if (!libcurl)
     		die("failed to load library '%s'", LIBCURL_FILE_NAME("libcurl"));
     
   -@@ compat/lazy-load-curl.c: CURLcode curl_global_init(long flags)
   +@@ compat/lazyload-curl.c: CURLcode curl_global_init(long flags)
     
     CURLsslset curl_global_sslset(curl_sslbackend id, const char *name, const curl_ssl_backend ***avail)
     {
   ``````

* 4:  d6289e3e050a ! 4:  05efe31e8f85 mingw: do load libcurl dynamically by default

   ``````diff
   @@ Metadata
     ## Commit message ##
        mingw: do load libcurl dynamically by default
    
   -    This will help with Git for Windows' maintenance going forward.
   +    This will help with Git for Windows' maintenance going forward: It
   +    allows Git for Windows to switch its primary libcurl to a variant
   +    without the OpenSSL backend, while still loading an alternate when
   +    setting `http.sslBackend = openssl`.
   +
   +    This is necessary to avoid maintenance headaches with upgrading OpenSSL:
   +    its major version name is encoded in the shared library's file name and
   +    hence major version updates (temporarily) break libraries that are
   +    linked against the OpenSSL library.
    
        Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>
    
   ``````

</details>
